### ISSUE TITLE
Private repo artifacts not downloaded / added to classpath

### ISSUE THREAD

**Issue by francisdb** [lihaoyi/mill#185]
```scala
object mymodule extends SbtModule {

  def repositories = super.repositories ++ Seq(
    // login and pass are globally configured
    MavenRepository("https://nexushost/repository/maven-releases", authentication = Some(coursier.core.Authentication(nexusUser, nexusPassword)))
  )
  ...
  override def ivyDeps = Agg(
    ivy"com.foo.bar::bar:6.1.1"
  )
}
```

A coursier download happens
/Users/francisdb/.coursier/cache/v1/https/francisdb%40nexushost/repository contains the pom for `com.foo.bar::bar` but no other files

`mill show mymodule.compileClasspath` does not contain `com.foo.bar::bar`

on a side note, the resolution failed message mentions `~/.ivy2` but not `~/.coursier`

```
Resolution failed for 1 modules:
--------------------------------------------
  com.foo.bar:bar_2.12:6.1.1
	not found: /Users/francisdb/.ivy2/local/...
	not found: https://repo1.maven.org/maven2/...
	not found: https://nexushost/repository/maven-releases/...
```

**Comment by DavidGregory084** [comment]
I have noticed this problem in SBT sometimes when using Coursier to resolve artifacts. I haven't yet dug into where it's going wrong in Coursier.

I think the silent failure in Mill might be caused by this snippet in [mill.scalalib.Lib.scala](https://github.com/lihaoyi/mill/blob/0.1.3/scalalib/src/mill/scalalib/Lib.scala#L92):

```scala
      val localArtifacts: Seq[File] = scalaz.concurrent.Task
        .gatherUnordered(sourceOrJar.map(Cache.file(_).run))
        .unsafePerformSync
        .flatMap(_.toOption)
```
This makes the (IMO) reasonable assumption that the JARs will always be present if Coursier doesn't return any resolution errors, so the `.flatMap(_.toOption)` discards any `None`s. This is why the `compileClasspath` is missing the entries for these JARs.

**Comment by francisdb** [comment]
I'm also using Coursier for our sbt build and there there are no issues fetching these artifacts. Anything I can do to debug this?

**Comment by DavidGregory084** [comment]
Do you know which version of Coursier you are using in your SBT build? IIRC Coursier's cache is versioned. It might be worth checking whether this is resolved if you add a newer version of Coursier (e.g. 1.0.2) to the Mill build [here](https://github.com/lihaoyi/mill/blob/master/build.sc#L84) and follow the instructions in the README to build and run Mill from source.

If you want to debug Mill it looks like it uses an environment variable $JAVA_OPTS in its launcher so you could add the JVM options for remote debug:

 `export JAVA_OPTS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005"`

and then debug in IntelliJ

**Comment by francisdb** [comment]
My sbt build has addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0")
I cleared the coursier cache and all that is added is the pom file, the jars are missing

**Comment by francisdb** [comment]
> This makes the (IMO) reasonable assumption that the JARs will always be present if Coursier doesn't return any resolution errors, so the .flatMap(_.toOption) discards any Nones. This is why the compileClasspath is missing the entries for these JARs.

Swallowing errors is never a good idea

**Comment by lihaoyi** [comment]
@francisdb I don't have any convenient private repos to test with. Do you want to help send a PR to fix this? The relevant code is what david pointed to above

**Comment by francisdb** [comment]
@lihaoyi sure, will have a look somewhere next week

**Comment by francisdb** [comment]
Somehow my artifact is marked optional, when enabling optional artifacts in `Lib.scala` everything works as expected

```scala
resolution.artifacts(true)
```

maybe related: https://github.com/coursier/coursier/issues/787

**Comment by lihaoyi** [comment]
@francisdb seems like a similar solution as https://github.com/lihaoyi/Ammonite/pull/713/files might fix everything then

**Comment by francisdb** [comment]
Yes, saw that pr, but would like to first find out why my direct dependency is being marked "optional"

**Comment by francisdb** [comment]
@lihaoyi as a temporary workaround we could include optional dependencies in the classpath
```
resolution.artifacts(withOptional = true)
```
not sure we need the rest of https://github.com/lihaoyi/Ammonite/pull/713/files

**Comment by francisdb** [comment]
According to https://github.com/coursier/coursier/issues/802#issuecomment-371129567 we should be pulling in optional dependencies (and ignore not found ones)
